### PR TITLE
exec: do not leak session IDs on errors

### PR DESCRIPTION
**Pull request by giuseppe**
always cleanup the exec session when the command specified to the "exec" is not found.
    
Closes: https://github.com/containers/podman/issues/20392

#### Does this PR introduce a user-facing change?

```release-note
exec sessions are not leaked when the specified command does not exist
```

### CODE REVIEW

File: test/system/075-exec.bats [hunk 1]
```diff
@@ -166,4 +166,19 @@ load helpers
...
+    cid="$output"
+
+    for i in {1..3}; do
+        run_podman 127 exec $cid blahblah
+        run_podman 125 exec -d $cid blahblah
+    done
+
+    run_podman inspect $cid
+    inspect_data="$output"
+    run jq '.[].ExecIDs | length' <<< "$inspect_data"
```
Comment (Luap99) [inline test/system/075-exec.bats:180]: no need for jq just add `--format "{{.ExecIDs}}"` then you can match the string against `[]`
or if you really want to check by number use `--format "{{len .ExecIDs}}"`

Comment (giuseppe) [reply test/system/075-exec.bats]: good idea, I'll use that

File: libpod/container_exec.go [hunk 2]
```diff
@@ -759,11 +759,21 @@ func (c *Container) Exec(config *ExecConfig, streams *define.AttachStreams, resi
 // Exec emulates the old Libpod exec API, providing a single call to create,
 // run, and remove an exec session. Returns exit code and error. Exit code is
 // not guaranteed to be set sanely if error is not nil.
-func (c *Container) exec(config *ExecConfig, streams *define.AttachStreams, resizeChan <-chan resize.TerminalSize, isHealthcheck bool) (int, error) {
+func (c *Container) exec(config *ExecConfig, streams *define.AttachStreams, resizeChan <-chan resize.TerminalSize, isHealthcheck bool) (ExitCode int, Err error) {
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply libpod/container_exec.go:762]: The linter agrees with you

**Review by vrothberg — APPROVED** [review]
LGTM

File: libpod/container_exec.go [hunk 3]
```diff
@@ -759,11 +759,21 @@ func (c *Container) Exec(config *ExecConfig, streams *define.AttachStreams, resi
...
 	if err != nil {
 		return -1, err
 	}
+	defer func() {
+		if err := c.ExecRemove(sessionID, false); err != nil {
+			if errors.Is(err, define.ErrNoSuchExecSession) {
+				return
+			}
+			exitCode = -1
+			retErr = err
```
Comment (rhatdan) [inline libpod/container_exec.go:773]: Shouldn't this only happen if retErr==nil?

Otherwise we should logrus the new error and continue to return retErr.

Comment (giuseppe) [reply libpod/container_exec.go]: yes that is better, fixed

**Comment by edsantiago** [comment]
This has broken CI. All remote system tests are now failing on (what looks like) all PRs. @giuseppe please treat this as urgent.

**Comment by giuseppe** [comment]
not even sure how to fix it properly.  @mheon should we add a new function to the bindings to allow ExecRemove from the remote client?  I don't think the cleanup should be done automatically server-side on failures

**Comment by giuseppe** [comment]
ideally, a merge should be possible only when the CI is all green (or has a way to explicitly force it if needed).  A `/lgtm` alone should not bypass the CI causing these emergencies

**Comment by edsantiago** [comment]
Oh! That's it, I see now. It merged while CI was still running. I hadn't seen that.

Yes, "ideally". Nobody anywhere in the universe will disagree with you. Unfortunately this is not an ideal world [citation needed]. That's why you see so many slash-lgtm-slash-hold: because slash-lgtm, while CI is running, is catastrophic.

Thank you for noticing that that's what happened. I feel better now that I understand.

**Pull request merged**